network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -196,6 +197,57 @@ func (c *ControlMsgValidationInspector) Inspect(from peer.ID, rpc *pubsub.RPC) e
 	return nil
 }
 
+// inspectIWant inspects RPC iWant control messages. This func will sample the iWants and perform validation on each iWant in the sample.
+// Ensuring that the following are true:
+// - Each iWant corresponds to an iHave that was sent.
+// - Each topic in the iWant sample is a valid topic.
+// If the number of iWants that do not have a corresponding iHave exceed the configured threshold an error is returned.
+// Args:
+// - iWant: the list of iWant control messages.
+// Returns:
+// - DuplicateFoundErr: if there are any duplicate message ids found in across any of the iWants.
+// - IWantCacheMissThresholdErr: if the rate of cache misses exceeds the configured allowed threshold.
+// - error: if any error occurs while sampling the iWants
+func (c *ControlMsgValidationInspector) inspectIWant(iWants []*pubsub_pb.ControlIWant) error {
+	sampleSize := uint(10 * c.rpcTracker.LastHighestIHaveRPCSize())
+	if sampleSize == 0 || sampleSize > c.config.IWantRPCInspectionConfig.MaxSampleSize {
+		sampleSize = c.config.IWantRPCInspectionConfig.MaxSampleSize
+	}
+
+	swap := func(i, j uint) {
+		iWants[i], iWants[j] = iWants[j], iWants[i]
+	}
+	err := c.sampleCtrlMessages(uint(len(iWants)), sampleSize, swap)
+	if err != nil {
+		return fmt.Errorf("failed to sample iwant messages: %w", err)
+	}
+
+	tracker := make(duplicateStrTracker)
+	cacheMisses := float64(0)
+	totalMessageIDS := float64(0)
+	for i := uint(0); i < sampleSize; i++ {
+		iWant := iWants[i]
+		for _, messageID := range iWant.GetMessageIDs() {
```

Comment:
For checking against duplicates, it is more efficient to build a map of `iWantPool[:sampleSize]`, and while building the map, check whether there is a duplicate. Using map than the nested `for` loops, we improve the time complexity from `O(sampleSize^2)` to `O(sampleSize)`.